src/ejabberd_c2s.erl

```diff
@@ -1133,29 +1135,21 @@ handle_incoming_message({send_filtered, Feature, From, To, Packet}, StateName, S
             ejabberd_router:route(From, To, FinalPacket),
             fsm_next_state(StateName, StateData)
     end;
-handle_incoming_message({broadcast, Type, From, Packet}, StateName, StateData) ->
-    %% this version is used only by mod_pubsub, which does things the old way
-    Recipients = mongoose_hooks:c2s_broadcast_recipients(StateData#state.server,
-                                                        [],
-                                                        StateData, Type, From, Packet),
-    lists:foreach(fun(USR) -> ejabberd_router:route(From, jid:make(USR), Packet) end,
-        lists:usort(Recipients)),
-    fsm_next_state(StateName, StateData);
-handle_incoming_message({call_info_handler, Tag, Data}, StateName, StateData) ->
-    NStateData =
-    case ejabberd_c2s_info_handler:get_for(Tag, StateData) of
-        {Tag, HandlerState} ->
-            case ejabberd_c2s_info_handler:safe_call(Tag, Data, HandlerState, StateData) of
-                {error, _} ->
-                    StateData;
-                NStateData0 ->
-                    NStateData0
-            end;
-        no_handler ->
-            ?INFO_MSG("event=no_info_handler tag=~p host=~p", [Tag, ejabberd_c2s_state:server(StateData)]),
-            StateData
-    end,
-    fsm_next_state(StateName, NStateData);
+handle_incoming_message({call_remote_hook, Tag, Args}, StateName, StateData) ->
+    Host = ejabberd_c2s_state:server(StateData),
+    HandlerState = maps:get(Tag, StateData#state.handlers, empty_state),
```

Comment:
I understand that `HandlerState` can be changed by the handler while the rest of the c2s state cannot. However, are we sure that there is no case where two handlers need to modify the same c2s data?
- If there is no such case, I think we could keep this approach.
- If we find one, I would have only one map called `HandlerState` which would be shared between the handlers (actually I would use this solution as the default here).

Reply:
I was thinking of providing the hook handlers with state info and how to make it compact yet convenient. One thing I'd like to avoid is breaking traceability - I mean that it's very poor experience when we enable tracing and each single `call` trace occupies 2 screens because of a huge `#state{}` passed in and out of a function.

My idea is as follows - let's make the handler name and state independent of each other. C2S state parts would be tagged (for example with namespaces), while each remote hook call would require from the caller to declare which parts of state it's interested in. For example, delegating a hook call to C2S would be:

```erlang
ejabberd_c2s:run_hook(C2SPid, Hook, StateTags, Args) ->
    C2SPid ! {run_hook, Hook, StateTags, Args}.
```

The dispatch in C2S would take a subset of the state:

```erlang
handle_info({run_hook, Hook, StateTags, Args}, State) ->
    Server = ...,
    Extensions = State#state.extensions,
    NewExtensions = mongoose_hooks:Hook(Server, [maps:with(StateTags, Extensions) | Args]),
    NewState = State#state{extensions = maps:merge(Extensions, NewExtensions)}
    ...
```

Where `#state.extensions` would be `#{xmlns() := term()}`, i.e. a map keyed with extension namespaces. The call site would declare which extensions it requires to read/write (not read-only variant, but also not the whole state is passed in):

```erlang
# e.g. mod_privacy
some_fun() ->
    ejabberd_c2s:run_hook(SessionPid, set_active_list, [?NS_PRIVACY])
```

This does not prevent handlers from writing to the state subset they get, but it makes it explicit which parts they get so it's easier to keep track of who changes what.

BTW, `StateTags` do not have to be XMPP namespaces, but since the handlers are usually coupled with XEPs, they seem to fit the purpose well.

Reply:
@erszcz The whole state record is passed to the handlers because some of them need to read the fields from the c2s state record. We decided at the tech meeting that refactoring it to a map is out of scope right now.

Your proposal basically falls between my two suggestion I presented at the tech meeting:
- Allowing any module to call an arbitrary MFA (here we allow F and A with fixed M = `mongoose_hooks`). We decided that this can bring too much chaos and increase the number of called functions (here: hooks) out of control.
- Having only one hook, which is implemented here.

Reply:
> We decided at the tech meeting that refactoring it to a map is out of scope right now.

I'm not suggesting to refactor it completely, but pushing the fields which are clearly coupled with extensions (roster, privacy) could make sense.

> Having only one hook, which is implemented here. I like this one solution better as the whole point of a hook is to be able to attach multiple handlers to it. We should not introduce a new hook (like set_active_list in your example) whenever there is a need for a new handler.

We're not on the same page here. `set_active_list` is just an example, I'm **not** for introducing such a new specific hook. This was only intended to illustrate the concept.

There is no note from the tech meeting that a discussion on _one hook_ vs _arbitrary MFA_ has taken place.

In any case, if we're just into having one hook as of now, we can still make it accept multiple tags to fragment state into parts and allow different handlers to get different parts of state. Food for thought.

Reply:
> I understand that `HandlerState` can be changed by the handler while the rest of the c2s state cannot. However, are we sure that there is no case where two handlers need to modify the same c2s data?

This seems easy - if you are handler 1, then you can simply call a function in the other module, since they operate on the same state and you can return the changed state. It gets more difficult when you have two handlers which are supposed to modify their separate states.

What we currently have is a hook which is designed to trigger one handler at a time, so if one handler decides it has to call another one which may change its state it has to issue another "remote" call to `self()`.

Reply:
@bartekgorny What I meant was that two different modules might want to modify e.g. roster subscriptions. Which means that it is not clear, which of them should "own" these fields. If we just share the mutable state between all handlers (have one map shared between them), there would be no such issue.